### DISCUSSION TITLE
Designing the disabled state of a composited widget

### DISCUSSION THREAD

**Discussion by pieterdd** [lapce/floem#228] [Q&A] [answered]
I'm developing a simple UI component library for Floem. Quick demo:

![Quick demo](https://github.com/pieterdd/prettygooey/raw/floem-rewrite/docs/img/showcase.png)

One of my current roadblocks is handling of disabled state. My [labeled checkbox](https://github.com/pieterdd/prettygooey/blob/floem-rewrite/src/checkbox.rs#L31) is, much like floem's [labeled checkbox](https://github.com/lapce/floem/blob/main/src/widgets/checkbox.rs#L28), a composite. But while the disabled state seems to propagate properly to the child components in the floem implementation, it doesn't in mine.

What should I do to have the disabled state propagate down to child components?

**Comment by dzhou121** [comment]
First of all, this looks really cool!

The disabled state isn't passed to child components though in Floem. Maybe it should?

**Reply by pieterdd** [reply]
Thanks! I like what you're doing to make it a lot easier for Rust apps to implement a native GUI.

I'd like to understand Floem's current operating model. If the disabled state isn't passed to child components, then how do the individual parts of Floem's labeled checkbox know to apply the grey style? For reference:

![image](https://github.com/lapce/floem/assets/1407980/8afade76-1b5d-4e72-9523-3112cb8727c7)

I see that Floem's labeled checkbox is an [`h_stack`](https://github.com/lapce/floem/blob/ac3654547948b600bc00ea62ce91fe4498d24f2f/src/widgets/checkbox.rs#L32). The stack contains an `svg` and a `label`, both of which seem to look different when the checkbox [is disabled](https://github.com/lapce/floem/blob/ac3654547948b600bc00ea62ce91fe4498d24f2f/examples/widget-gallery/src/checkbox.rs#L43). How do the `svg` and `label` know that they're supposed to render the disabled state if the `h_stack` isn't passing its state to its children?

**Reply by dzhou121** [reply]
I think it's because the h_stack applies the disabled style when it's disabled, and certain styles (e.g. font color) can be passed to the children.

**Reply by pieterdd** [reply]
Ah, that could indeed explain the font color. And what about the grey checkbox background? I don't have the impression that h_stack is setting a grey background, so I don't immediately see how the box could inherit it from its parent.

Based on the information so far I'm inclined to agree with the notion of having a disabled state propagate to children the same way font color already does. I might try to create a pull request for Floem, but I'm very new to all of this so I'm not sure I'll be able to help.

**Reply by jrmoulton** [reply]
https://github.com/lapce/floem/blob/ac3654547948b600bc00ea62ce91fe4498d24f2f/src/widgets/mod.rs#L88C4-L121C37

This is where the styling for the checkboxes is done

**Reply by Zoxc** [reply]
The default style sheet has a separate style for `CheckboxClass` inside a disabled `LabeledCheckboxClass`. No inheritance is needed.

**Reply by pieterdd** [reply]
Very interesting you bring that up - I might have misunderstood how the class system works. But it looks promising. Are my assumptions correct now?

1. Child components that should be restyled upon disable should listen to a class, e.g. `child_component.class(ChildComponentClass)`
2. The parent component's style should overwrite the child's style upon disable:

```rust
parent_component.style(|s| {
  s.disabled(|s| {
    s.class(ChildComponentClass, |s| /* set disabled style for the child component here */)
  })
})
```

I tried applying this in a minimal example but unfortunately it doesn't respond how I expected:

```rust
use floem::peniko::Color;
use floem::reactive::{create_signal, ReadSignal, WriteSignal};
use floem::style::{AlignItems, Style};
use floem::view::View;
use floem::views::{h_stack, label, v_stack, Decorators};
use floem::widgets::{button, checkbox};
use floem::{style_class, EventPropagation};

style_class!(pub ButtonStyle);

fn spinbox(counter: ReadSignal<i32>, set_counter: WriteSignal<i32>) -> impl View {
    h_stack((
        button(|| "-").class(ButtonStyle).on_click(move |_| {
            set_counter.update(|value| *value -= 1);
            EventPropagation::Stop
        }),
        label(move || counter.get()),
        button(|| "+").class(ButtonStyle).on_click(move |_| {
            set_counter.update(|value| *value += 1);
            EventPropagation::Stop
        }),
    ))
    .style(|s| {
        let button_style = Style::new()
            .background(Color::GREEN)
            .color(Color::WHITE)
            .font_size(18.0)
            .padding(5.0);

        s.class(ButtonStyle, move |_| button_style.clone())
            .align_items(AlignItems::Center)
            .disabled(|s| {
                s.background(Color::DIM_GRAY)
                    .class(ButtonStyle, |s| s.background(Color::DARK_GRAY))
            })
            .gap(10.0, 0.0)
    })
}

fn app_view() -> impl View {
    let (counter, set_counter) = create_signal(0);
    let (spinbox_enabled, set_spinbox_enabled) = create_signal(true);

    v_stack((
        checkbox(spinbox_enabled)
            .on_click_stop(move |_| set_spinbox_enabled.set(!spinbox_enabled.get())),
        spinbox(counter, set_counter).disabled(move || !spinbox_enabled.get()),
    ))
}

fn main() {
    floem::launch(app_view);
}
```

![enabled](https://github.com/lapce/floem/assets/1407980/55a3a98b-c6d4-4080-b155-890d3f7d6a2a) ![disabled](https://github.com/lapce/floem/assets/1407980/ec266d4b-1f10-4c1d-81bc-75dea77ff04a)

How come the buttons aren't changing their background to dark gray?

**Reply by Zoxc** [reply]
That code looks correct skimming over it. Try with https://github.com/lapce/floem/pull/229.

**Reply by pieterdd** [reply]
Nice, that works!

![disabled state](https://github.com/lapce/floem/assets/1407980/f833b08c-c9f7-482c-a302-accf80eccbb7)

**Reply by pieterdd** [reply]
**(Edit: feel free to ignore the post below. Once https://github.com/lapce/floem/issues/230 is resolved I will no longer rely on classes to achieve disabled state in descendants.)**

I have another case regarding groups of widgets.

Individually, my widgets implement a disabled state. But under my current understanding of the system, if I create a group of these widgets, the group should re-implement the disabled state for every widget in the group.

In the example below, the top two widgets are ungrouped while the bottom two are grouped:

![enabled](https://github.com/lapce/floem/assets/1407980/a085df7c-399f-4130-abeb-3562d94cd9d7) ![disabled](https://github.com/lapce/floem/assets/1407980/841da7fd-4563-4b22-a32f-3c5c257fdc3a)

When the checkbox is unchecked, the ungrouped widgets correctly switch to the disabled state while the grouped ones don't. Am I using Floem correctly or is this perhaps a bug?

```rust
use floem::peniko::Color;
use floem::reactive::create_signal;
use floem::style::AlignItems;
use floem::style_class;
use floem::view::View;
use floem::views::{h_stack, label, svg, v_stack, Decorators};
use floem::widgets::checkbox;

style_class!(pub BulletStyle);

fn bullet_item(item_label: &'static str) -> impl View {
    h_stack((
        svg(|| String::from("")).class(BulletStyle),
        label(move || item_label),
    ))
    .style(|s| {
        s.align_items(AlignItems::Center)
            .class(BulletStyle, |s| {
                s.width(15.0)
                    .height(15.0)
                    .border_radius(100.0)
                    .background(Color::GREEN)
            })
            .disabled(|s| s.class(BulletStyle, |s| s.background(Color::GRAY)))
            .gap(10.0, 0.0)
    })
}

fn group_of_bullet_items() -> impl View {
    v_stack((
        bullet_item("Grouped bullet item 1"),
        bullet_item("Grouped bullet item 2"),
    ))
    .style(|s| {
        s.background(Color::rgb(0.8, 0.8, 0.8))
            .disabled(|s| s.class(BulletStyle, |s| s.background(Color::GRAY)))
    })
}

fn app_view() -> impl View {
    let (controls_enabled, set_controls_enabled) = create_signal(true);

    v_stack((
        checkbox(controls_enabled)
            .on_click_stop(move |_| set_controls_enabled.set(!controls_enabled.get())),
        bullet_item("Ungrouped bullet item 1").disabled(move || !controls_enabled.get()),
        bullet_item("Ungrouped bullet item 2").disabled(move || !controls_enabled.get()),
        group_of_bullet_items().disabled(move || !controls_enabled.get()),
    ))
}

fn main() {
    floem::launch(app_view);
}
```

**Comment by pieterdd** [comment, marked as answer]
Quick update to this thread in case someone finds it through a search engine later: as of e795021bfb28cd15a6d499349e547c435ceb8520, Floem ensures that views pass their disabled state down to its descendants for styling purposes. If a parent is disabled, its descendants will apply disabled styling.
